client/components/site-selector/index.jsx

```diff
@@ -295,6 +286,78 @@ class SiteSelector extends Component {
 			sites = sites.filter( site => site.slug !== this.props.selected );
 		}
 
+		return sites;
+	}
+
+	shouldRenderRecentSites() {
+		return (
+			this.props.showRecentSites &&
+			this.shouldShowGroups() &&
+			! this.props.sitesFound &&
+			size( this.props.recentSites )
+		);
+	}
+
+	renderAllSites() {
+		if ( ! this.props.showAllSites || this.props.sitesFound || ! this.props.allSitesPath ) {
+			return null;
+		}
+
+		this.visibleSites.push( ALL_SITES );
+
+		const isHighlighted = this.isHighlighted( ALL_SITES );
+
+		return (
+			<AllSites
+				key="selector-all-sites"
+				sites={ this.props.sites }
+				onSelect={ this.onAllSitesSelect }
+				onMouseEnter={ this.onAllSitesHover }
+				isHighlighted={ isHighlighted }
+				isSelected={ this.isSelected( ALL_SITES ) }
+			/>
+		);
+	}
+
+	renderRecentSites() {
+		if ( ! this.shouldRenderRecentSites() ) {
+			return null;
+		}
+
+		const sitesToBeRendered = this.sitesToBeRendered();
+
+		const sites = [];
+		for ( const siteId of this.props.recentSites ) {
+			const site = find( sitesToBeRendered, { ID: siteId } );
+			if ( site ) {
+				sites.push( site );
+			}
+		}
+
+		if ( ! sites ) {
+			return null;
+		}
+
+		const recentSites = sites.map( this.renderSite, this );
+
+		return <div className="site-selector__recent">{ recentSites }</div>;
+	}
+
+	renderSites() {
+		let sites;
+
+		if ( ! this.props.hasAllSitesList ) {
+			return <SitePlaceholder key="site-placeholder" />;
+		}
+
+		sites = this.sitesToBeRendered();
```

Comment:
It's unfortunate that `this.sitesToBeRendered` is called twice on every render. It's quite an expensive operation. Could be computed just once and passed down as param to `renderRecentSites` and `renderSites`?